File: LeetCode/LeetCode-CPP/2020JuneLeetCodingChallenge/Search in a Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
/*
class Solution {
public:
    TreeNode* res = NULL;
    void dfs(TreeNode* root, int val) {
        if(root == NULL)
            return;
        if(root->val == val)
            res = root;
        dfs(root->left, val);
        dfs(root->right, val);
    }
    
    TreeNode* searchBST(TreeNode* root, int val) {
        dfs(root, val);
        return res;
    }
};
*/

class Solution {
public:
        TreeNode* searchBST(TreeNode* root, int val) {
        if(root == NULL)
            return NULL;
        TreeNode* res = NULL;
        queue<TreeNode*> q;
        q.push(root);
        while(!q.empty()) {
            TreeNode* node = q.front();
            q.pop();
            if(node->val == val) {
                return node;
            }
            if(node->left)
                q.push(node->left);
            if(node->right)
                q.push(node->right);
        }
        return res;
    }
};

File: LeetCode/LeetCode-Java/2020JuneLeetCodingChallenge/Search in a Binary Search Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
// class Solution {
//     public TreeNode searchBST(TreeNode root, int val) {
//         if(root == null) {
//             return null;
//         }
//         if(val < root.val) {
//             return searchBST(root.left, val);
//         } else if(val > root.val) {
//             return searchBST(root.right, val);
//         }
//         return root;
//     }
// }

class Solution {
    public TreeNode searchBST(TreeNode root, int val) {
        if(root == null) {
            return null;
        }
        TreeNode ans = root;
        
        while(ans != null) { // 数据不为空
            if(ans.val > val) {
                ans = ans.left;
            } else if(ans.val < val) {
                ans = ans.right;
            } else {
                return ans;
            }
        }
        return ans;
    }
}